BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m

```diff
@@ -115,17 +164,49 @@ - (NSURLSession *)createBackgroundSession
         queue.maxConcurrentOperationCount = 8;
 
         _backgroundSession = [NSURLSession sessionWithConfiguration:sessionConfig delegate:self delegateQueue:queue];
-        self.backgroundSessionIdToSessionTaskIdAndTaskDelegate[backgroundSessionIdentifier] = [NSMapTable strongToWeakObjectsMapTable];
+        self.backgroundSessionIdToSessionTask[backgroundSessionIdentifier] = [NSMapTable strongToWeakObjectsMapTable];
     }
     return _backgroundSession;
 }
 
+- (BOXURLSessionCacheClient *)createCacheClient:(NSString *)rootCacheDir delegate:(id<BOXURLSessionCacheClientDelegate>)delegate
+{
+    if (_cacheClient == nil) {
+        _cacheClient = [[BOXURLSessionCacheClient alloc] initWithCacheRootDir:rootCacheDir];
+        _cacheClient.delegate = delegate;
+    }
+    return _cacheClient;
+}
+
 #pragma mark - public methods
 
-- (void)setUpToSupportBackgroundTasksWithDefaultDelegate:(id<BOXURLSessionManagerDelegate>)delegate
+- (void)setUpToSupportBackgroundTasksWithDefaultDelegate:(id<BOXURLSessionManagerDelegate>)delegate rootCacheDir:(NSString *)rootCacheDir
 {
     self.defaultDelegate = delegate;
     [self createBackgroundSession];
+
+    //prepare the map of background session id to its NSURLSession object, which is needed to support background sessions created from extensions
+    if (_backgroundSessionIdToSession == nil) {
+        _backgroundSessionIdToSession = [NSMutableDictionary new];
+    }
+    self.backgroundSessionIdToSession[backgroundSessionIdentifier] = self.backgroundSession;
+
+    [self createCacheClient:rootCacheDir delegate:delegate];
+
+    //populate pending background session tasks
+    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+
+    [self pendingBackgroundDownloadUploadSessionTasks:^(NSArray<NSURLSessionUploadTask *> * _Nonnull uploadTasks, NSArray<NSURLSessionDownloadTask *> * _Nonnull downloadTasks) {
+        for (NSURLSessionUploadTask *uploadTask in uploadTasks) {
+            [self associateBackgroundSessionTask:uploadTask withTaskDelegate:nil];
+        }
+        for (NSURLSessionDownloadTask *downloadTask in downloadTasks) {
+            [self associateBackgroundSessionTask:downloadTask withTaskDelegate:nil];
+        }
+        dispatch_semaphore_signal(sema);
+    }];
+
+    dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
```

Comment:
The DISPATCH_TIME_FOREVER feels unsafe. Can we set a timeout here and a retry mechanism?